src/server/pps/server/api_server.go

```diff
@@ -2148,37 +2161,90 @@ func (a *apiServer) CreatePipeline(ctx context.Context, request *pps.CreatePipel
 	metricsFn := metrics.ReportUserAction(ctx, a.reporter, "CreatePipeline")
 	defer func(start time.Time) { metricsFn(start, retErr) }(time.Now())
 
+	js, err := protojson.Marshal(request)
+	if err != nil {
+		return nil, unknownError(ctx, "could not marshal CreatePipelineRequest to JSON", err)
+	}
+	v2Req := &pps.CreatePipelineV2Request{
+		CreatePipelineRequestJson: string(js),
+		Update:                    request.Update,
+		Reprocess:                 request.Reprocess,
+	}
+	if _, err := a.createPipeline(ctx, v2Req); err != nil {
+		return nil, err
+	}
+	return &emptypb.Empty{}, nil
+}
+
+func (a *apiServer) createPipeline(ctx context.Context, req *pps.CreatePipelineV2Request) (effectiveSpecJSON string, err error) {
+	clusterDefaultsResponse, err := a.GetClusterDefaults(ctx, &pps.GetClusterDefaultsRequest{})
+	if err != nil {
+		return "", status.Error(codes.Internal, "could not get cluster defaults")
+	}
+
+	defaultsJSON := clusterDefaultsResponse.GetClusterDefaultsJson()
+	if defaultsJSON == "" {
+		defaultsJSON = "{}"
+	}
+	reqJSON, err := json.Marshal(struct {
```

Comment:
I wonder if we actually want to make ClusterDefaultsJson handle adding the `{"create_pipeline_request":...}` wrapping?  It would kind of make sense that maybe ClusterDefaultsJson has defaults other than `create_pipeline_request`.

It would be nice to guarantee that the next few lines don't error.  It's unlikely, but if you can enter this state, it would be best to reject it at the time when the defaults that cause this behavior are set.

Reply:
> I wonder if we actually want to make ClusterDefaultsJson handle adding the `{"create_pipeline_request":...} ` wrapping?

How so? What would that look like?

> It would kind of make sense that maybe ClusterDefaultsJson has defaults other than `create_pipeline_request`.

It definitely could, although in this case only `create_pipeline_request` matters.

> It would be nice to guarantee that the next few lines don't error.

They always will be able to error if the user submits invalid input. _However_, `SetClusterDefaults` could validate that `create_pipeline_request` is valid. I’ll add that.